src/Qir/Runtime/public/QubitManager.hpp

```diff
@@ -0,0 +1,220 @@
+
+    // QubitListInSharedArray implements a singly-linked list with "pointers" to the first and the last element stored.
+    // Pointers are the indexes in a shared array. Shared array isn't sotored here because it can be reallocated.
+    // This class maintains status of elements in the list by virtue of linking them as part of this list.
+    // This class doesn't update status of elementes excluded from the list.
+    // This class is small, contains no C++ pointers and relies on default shallow copying/destruction.
+    struct QubitListInSharedArray final
+    {
+    private:
+        QubitIdType firstElement = NoneMarker;
+        QubitIdType lastElement = NoneMarker;
+        // We are not storing pointer to shared array because it can be reallocated.
+        // Indexes and special values remain the same on such reallocations.
+
+    public:
+        // Initialize empty list
+        QubitListInSharedArray() = default;
+
+        // Initialize as a list with sequential elements from startId to endId inclusve.
+        QubitListInSharedArray(QubitIdType startId, QubitIdType endId, QubitIdType* sharedQubitStatusArray);
+
+        bool IsEmpty() const;
+        void AddQubit(QubitIdType id, bool addToFront, QubitIdType* sharedQubitStatusArray);
+        int TakeQubitFromFront(QubitIdType* sharedQubitStatusArray);
+        void MoveAllQubitsFrom(QubitListInSharedArray& source, QubitIdType* sharedQubitStatusArray);
+    };
+
+
+    // Restricted reuse area consists of multiple segments. Qubits released in one segment cannot be reused in another.
+    // One restricted reuse area can be nested in a segment of another restricted reuse area.
+    // This class tracks current segment of an area. Previous segments are tracked collectively (not individually).
+    // This class is small, contains no C++ pointers and relies on default shallow copying/destruction.
+    struct RestrictedReuseArea final
+    {
+    public:
+        QubitListInSharedArray FreeQubitsReuseProhibited;
+        QubitListInSharedArray FreeQubitsReuseAllowed;
+
+        RestrictedReuseArea() = default;
+        RestrictedReuseArea(QubitListInSharedArray freeQubits);
+    };
+
+
+
+    // This is NOT a stack! We modify it only by push/pop, but we also iterate over elements.
+    // TODO: Better name?
+    class CRestrictedReuseAreaStack final : public std::vector<RestrictedReuseArea>
+    {
+    public:
+        // No complex scenarios for now. Don't need to support copying/moving.
+        CRestrictedReuseAreaStack() = default;
+        CRestrictedReuseAreaStack(const CRestrictedReuseAreaStack&) = delete;
+        CRestrictedReuseAreaStack& operator = (const CRestrictedReuseAreaStack&) = delete;
+        ~CRestrictedReuseAreaStack() = default;
+
+        void PushToBack(RestrictedReuseArea area);
+        RestrictedReuseArea PopFromBack();
+        RestrictedReuseArea& PeekBack();
+        // TODO: Remove and use size directly?
+        int Count() const;
+    };
```

Comment:
(Up to you)
Consider naming the member functions the same way as in [`std::stack`](https://www.cplusplus.com/reference/stack/stack/?kw=stack): e.g. `Push()`, `Pop()`, `Top()`, `Size()` (and referring to the documentation of the `std::stack`).

Reply:
This is not a pure stack. While Push and Pop are only done only LIFO style, we also iterate it back and forth. I'll keep the names. I did switch to std iterator for iteration.